src/marina_bucket.erl

```diff
@@ -0,0 +1,207 @@
+    gen_server:call(Name, stop).
+
+%% @private
+%% @doc Handling call messages
+-spec(handle_call(Request :: term(), From :: {pid(), Tag :: term()},
+    State :: #{}) ->
+    {reply, Reply :: term(), NewState :: #{}} |
+    {reply, Reply :: term(), NewState :: #{}, timeout() | hibernate} |
+    {noreply, NewState :: #{}} |
+    {noreply, NewState :: #{}, timeout() | hibernate} |
+    {stop, Reason :: term(), Reply :: term(), NewState :: #{}} |
+    {stop, Reason :: term(), NewState :: #{}}).
+
+handle_call(increase_max_ticket, _, State = #{max_ticket := Max}) when Max < ?STOP_POLLING_THRESHOLD ->
+    schedule_poll(self()),
+    {reply, ok, State#{max_ticket => Max + 1}};
+handle_call(increase_max_ticket, _, State) ->
+    {reply, ignored, State#{tref => undefined}};
+%% too many tickets checkedout, only allow return, don't allow accquire any more.
+handle_call(acquire_ticket, {From, _Ref}, State = #{checkedout_ticket := T, max_ticket := Max, name := Pool}) when T >= Max ->
+    lager:warning("cannot aacquire tickets!! pool ~p from ~p, waiting for all tickets to be returned, checkedout ~p, max ~p", [Pool, From, T, Max]),
+    {reply, error, State};
+%% a pid acquires a ticket, give the ticket to the process and descrease the total number of etickets;
+handle_call(acquire_ticket, {From, _Ref}, State) ->
+    State1 = checkout_ticket(From, State),
+    {reply, ok, State1};
+handle_call({return_ticket, Reason}, {From, _Ref}, State = #{name := Pool, checkedout_ticket := Tickets, max_ticket := Max, bucket := Dict}) ->
+    lager:debug("release ticket from pool ~p, current tickets: ~p", [Pool, Tickets]),
+    {Dict1, TicketsToReturn} =
+        case dict:find(From, Dict) of
+            error ->
+                %% some process going to return a ticket but it is not recorded, there must be a leak
+                %% TODO: fix possible leak
+                {Dict, 0};
+            {ok, 1} ->
+                {dict:erase(From, Dict), 1};
+            {ok, N} when N > 1 ->
+                {dict:store(From, N - 1, Dict), 1}
+        end,
+    NewState = adjust_max_tickets(Reason, State),
+    {reply, ok, NewState#{checkedout_ticket => (Tickets - TicketsToReturn), bucket => Dict1}};
+handle_call(stop, _From, State) ->
+    {stop, normal, State}.
+
+%% @private
+%% @doc Handling cast messages
+-spec(handle_cast(Request :: term(), State :: #{}) ->
+    {noreply, NewState :: #{}} |
+    {noreply, NewState :: #{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #{}}).
+handle_cast(_Request, State = #{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc Handling all non call/cast messages
+-spec(handle_info(Info :: timeout() | term(), State :: #{}) ->
+    {noreply, NewState :: #{}} |
+    {noreply, NewState :: #{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #{}}).
+handle_info({'DOWN', _MonitorRef, process, From, _Info}, State = #{name := Pool, checkedout_ticket := Tickets, bucket := Dict}) ->
```

Comment:
where does this "DOWN" come from? from marina worker? 'DOWN' message is received from terminated process when you monitor it. But it seems we just link processes (by gen_server:start_link),
if we want to trap the exis signal, the message would be "EXIT" and need to do process_flag(trap_exit, true)

Reply:
The caller, when the caller is down from some reason, we need to release possible ticket from our bucket.